Support Arnold 7 and 6.2 in a single build

johnhaddon opened the pull request:
This moves IECoreArnold from Cortex into Gaffer, fixes a few Arnold 7 issues, and adds dual Arnold 7/6.2 support to Gaffer. The intention is that for Gaffer 0.61, a single Gaffer download will serve both Arnold 6.2 and 7, choosing the appropriate version at runtime.

danieldresser-ie reviewed (commented):
LGTM.  I also checked with Andrew, and he doesn't think his hesitations about this organization are worth holding this up for.

johnhaddon commented on python/GafferArnold/__init__.py:
```diff
@@ -42,18 +42,16 @@
-	# _without_ RTLD_GLOBAL. This prevents clashes between the
-	# LLVM symbols in libai.so and the Mesa OpenGL driver.
-	# Ideally we wouldn't use RTLD_GLOBAL anywhere - see
+	# Make sure we _GafferArnold _without_ RTLD_GLOBAL. This prevents clashes
```
> I'm assuming the reason why we no longer do this to IECoreArnold here is because as part of moving it from Cortex, we no longer compile it with RTLD_GLOBAL?

Ah, no, this was an error actually. It made sense to remove it in the earlier commit because `_IECoreArnold` no longer existed (it was just additional bindings in `_GafferArnold`). But not it is a standalone module, it needs to come back. I'll sort it out. I also need to get to the bottom of these Mac failures, and I think I can improve things by taking a more principled approach to the problem of the Arnold log and profile still being a global resource.

As an aside, RTLD_GLOBAL is a runtime thing, not a compile time thing. It gets turned on in `IECore/__init__.py` to work around historical problems, and I'm keen to remove it completely.

andrewkaufman replied:
Out of curiosity, what is holding us back from removing it? At IE we flipped it off back in March and haven’t had any issues.

johnhaddon replied:
That's documented here : https://github.com/GafferHQ/gaffer/pull/4175. Problem with Appleseed, which is presumably why it hasn't affected IE?

johnhaddon replied:
I fixed this in place here : https://github.com/GafferHQ/gaffer/pull/4461/commits/33d00d5b4f9fe04de89e67341b406d90352bbdec#diff-0de7066c8689f99f9be21d618e552ba306211985159e7b5f4b2d5f8277ea5a78R37-R50.

johnhaddon commented:
I've rebased this to include the `RTLD_GLOBAL` change discussed inline, and added 3 further commits on the end. All tests are now passing on Linux, but we have remaining failures on Mac for Arnold 7. I've whittled those down to a very small standalone repro using only the Arnold Python API, and provided it to SolidAngle. I'm hoping they can either point to some misuse of the API or fix on their side.

In the meantime, the question is "what do we do?". Regardless of the Mac failures, I think there's a lot of value in making Arnold 7 available for beta testing, so I'm somewhat inclined to merge this anyway (assuming the new commits look OK). Any objections or other ideas?

danieldresser-ie commented:
All looks reasonable, and I agree with disabling the mac test in order to get a release out.

johnhaddon commented:
> I agree with disabling the mac test in order to get a release out.

Done in https://github.com/GafferHQ/gaffer/pull/4461/commits/082062be44d204a5812e5a4cdc65c317f4d90366. I've also been able to confirm that the tests pass in the as-yet-unreleased 7.0.0.1.
